Felienne/hedy
# [CLEAN-UP] Improved coding for loading programs by id

**Description**
In this PR we clean-up the old structure on which we retrieved programs by their id. Currently we use a code to differentiate between a normal step and a program id, namely this:

`if step and isinstance(step, str) and len(step) > 2:`

The issue is: we don't use steps! This might be some relic code from supporting sub-levels, but the step parameters is completely redundant. So currently it is only used to store the program id when retrieving one, making this current code a bit weird. We clean-up the code and replace step with `program_id` and remove this validation line. We also solve another important bug: **A teacher was able to see all programs, regardless if they where from their own students or not**. This bug occurred due to only checking if someone was a teacher, but not if it was the teacher of the owner of the program. 

**Fixes**
This PR fixes #2117

**How to test**
This is mostly a clean-up. One think to verify is that as a teacher you are not able to visit a private program of a user that is not in your class. Als verify that the public program viewing on the code page (not the dedicated viewing page) now has a new url.

**Checklist**
Done? Check if you have it all in place using this list*
  

* If you're unsure about any of these, don't hesitate to ask. We're here to help!
